Fixes #1329, #1328. Add a CSP reporting endpoint and move security headers into app.

miketaylr opened the pull request:
This will report CSP violations to `/tmp/webcompat-csp-reports.log` (and DevTools also will dump them for the page you're on). Let's let this run wild for a week or so then figure out what needs to change. 

(I already noticed the following, which points to lodash):
```
[Report Only] Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' https://www.google-analytics.com".
```

r? @karlcow

miketaylr commented on webcompat/views.py:
```diff
@@ -306,3 +321,13 @@ def contributors():
+        return
+    with open(app.config['CSP_REPORTS_LOG'], 'a') as r:
+        r.write(request.data + '\n')
+    return ('', 204)
```
I don't know if `204` is the best here. ❓

miketaylr commented:
(just pushed 2 simple tests and a return value for the non-correct content-type case)

karlcow commented on webcompat/views.py:
```diff
@@ -306,3 +321,13 @@ def contributors():
+
+
+@app.route('/csp-report', methods=['POST'])
+def log_report():
```
nit: maybe log_csp_report()

miketaylr replied:
done!

karlcow commented on webcompat/views.py:
```diff
@@ -306,3 +321,13 @@ def contributors():
+@app.route('/csp-report', methods=['POST'])
+def log_report():
+    '''Route to record CSP header violations.'''
+    if 'application/csp-report' not in request.headers.get('content-type'):
```
```python
request.headers.get('content-type', '')
```

to avoid people trying to hit the server with wrong request and generating 500.

miketaylr replied:
good point, done.

karlcow commented on webcompat/views.py:
```diff
@@ -306,3 +321,13 @@ def contributors():
+    if 'application/csp-report' not in request.headers.get('content-type'):
+        return ('Wrong Content-Type.', 400)
+    with open(app.config['CSP_REPORTS_LOG'], 'a') as r:
+        r.write(request.data + '\n')
```
nothing weird but you could have used the logging module too which we instantiate.

karlcow replied:
Another reason to do that is with the combination of a logging file configuration to allow for activating and deactivating the feature when needed and have timestamp when the log is happening. 

```python
CSP_LOG = True # or False
if CSP_LOG:
    csp_log.info(request.data)
```

karlcow replied:
when CSP_LOG is False. You can just send a [403](https://tools.ietf.org/html/rfc7231#section-6.5.3).

miketaylr replied:
Good point, let's do this.

miketaylr replied:
@karlcow one question:

Currently the default `file_handler` is set to `logging.INFO` level, which means anything `INFO` and above will get logged. For CSP, probably something like `logging.WARNING` is appropriate... but that means CSP logs will pollute the regular `webcompat.log`. Which I think we don't want.

The current dumb approach writes to a its own log file, so it avoids this problem.

Is there something obvious I'm missing? Admittedly, the logging module is confusing. 😕

karlcow replied:
So my suggestion was a bad idea, given the context of flask which is exploding the normal logging python module. The [logging module](https://docs.python.org/2/library/logging.html) is ["simple"](https://pymotw.com/3/logging/index.html), but flask makes it more difficult. Cf. https://github.com/pallets/flask/issues/2023

so let's go with your initial suggestion.

karlcow commented on webcompat/views.py:
```diff
@@ -51,6 +51,21 @@ def before_request():
 @app.after_request
 def after_request(response):
     session_db.remove()
+    # Security headers
```
I would put the full csp thing in a specific function: `add_csp(response)` which returns `response` with the new headers, that I would call here.

miketaylr replied:
Thanks, agree that's less messy. I'll define add_csp in helpers.py.

miketaylr replied:
...and add_sec_headers.

karlcow commented on webcompat/views.py:
```diff
@@ -306,3 +321,13 @@ def contributors():
+        return ('Wrong Content-Type.', 400)
+    with open(app.config['CSP_REPORTS_LOG'], 'a') as r:
+        r.write(request.data + '\n')
+    return ('', 204)
```
[204](https://tools.ietf.org/html/rfc7231#section-6.3.5) means 'No content' and is indeed the appropriate response.

karlcow commented on tests/test_urls.py:
```diff
@@ -93,6 +93,17 @@ def test_labeler_webhook(self):
         # A random post should 401, only requests from GitHub will 200
         self.assertEqual(rv.status_code, 401)
 
+    def test_csp_report_uri(self):
```
Missing comment, explaining the test.

miketaylr replied:
oops!

miketaylr replied:
done.

karlcow commented on tests/test_urls.py:
```diff
@@ -93,6 +93,17 @@ def test_labeler_webhook(self):
+        rv = self.app.post('/csp-report', headers=headers)
+        self.assertEqual(rv.status_code, 204)
+
+    def test_csp_report_uri_bad_content_type(self):
```
missing comment explaining the test.

miketaylr replied:
done.

karlcow commented:
nosetests -v 

```
…
test_csp_report_uri (tests.test_urls.TestURLs) ... ok
test_csp_report_uri_bad_content_type (tests.test_urls.TestURLs) ... ok
…
```

miketaylr commented:
Have addressed all feedback except the CSP_LOG part -- will get to that soon.

> This branch has conflicts that must be resolved

Looks like I need to rebase and fix conflicts as well.

miketaylr commented:
@karlcow ok, let's go back to the original log file.

r? for https://github.com/webcompat/webcompat.com/pull/1367/commits/c8065f7b0594e84ed7aa795b5437638c378d1359

karlcow commented:
ok r+ @miketaylr

miketaylr commented:
Thanks @karlcow.
